### PR TITLE
nft requests

### PR DESCRIPTION
**Pull request by Joeysantoro**

### CODE REVIEW

File: src/interfaces/IERC7540.sol [hunk 1]
```diff
@@ -22,6 +22,18 @@ interface IERC165 {
     function supportsInterface(bytes4 interfaceId) external view returns (bool);
 }
 
+interface IERC7540Base {
+    // Required
+    function ownerOf(uint256 rid);
```
Comment (0xfarhaan) [inline src/interfaces/IERC7540.sol:27]:
```suggestion
    function ownerOfRedemptionRequest(uint256 rid);
```

File: src/interfaces/IERC7540.sol [hunk 2]
```diff
@@ -22,6 +22,18 @@ interface IERC165 {
...
+
+    function claimRequest(uint256 rid, address to);
```
Comment (0xfarhaan) [inline src/interfaces/IERC7540.sol:29]:
```suggestion
    function claimRedemptionRequest(uint256 rid, address to);
```

File: src/interfaces/IERC7540.sol [hunk 3]
```diff
@@ -22,6 +22,18 @@ interface IERC165 {
...
+
+    // Not strictly required
+    function creatorOf(uint256 rid);
```
Comment (0xfarhaan) [inline src/interfaces/IERC7540.sol:32]:
```suggestion
    function creatorOfRedemptionRequest(uint256 rid);
```

File: src/interfaces/IERC7540.sol [hunk 4]
```diff
@@ -22,6 +22,18 @@ interface IERC165 {
...
+
+    function transferRequest(uint256 rid, address to);
```
Comment (0xfarhaan) [inline src/interfaces/IERC7540.sol:34]:
```suggestion
    function transferRedemptionRequest(uint256 rid, address to);
```

File: src/interfaces/IERC7540.sol (hunk 1 above)
Comment (0xfarhaan) [reply src/interfaces/IERC7540.sol:27]: and then the equivalent for deposit for each

File: src/interfaces/IERC7540.sol [hunk 5]
```diff
@@ -22,18 +22,65 @@ interface IERC165 {
     function supportsInterface(bytes4 interfaceId) external view returns (bool);
 }
 
+inerface IERC7540DepositReceiver {
```
Comment (Joeysantoro) [inline src/interfaces/IERC7540.sol:25]:
```suggestion
interface IERC7540DepositReceiver {
```

File: src/interfaces/IERC7540.sol [hunk 6]
```diff
@@ -64,12 +111,18 @@ interface IERC7540Redeem {
      * - MUST NOT show any variations depending on the caller.
      * - MUST NOT revert unless due to integer overflow caused by an unreasonably large input.
      */
-    function pendingRedeemRequest(address operator) external view returns (uint256 shares);
+    function pendingRedeemRequest(uint256 rid) external view returns (uint256 shares);
```
Comment (0xfarhaan) [inline src/interfaces/IERC7540.sol:114]: Missing transferRedeemRequest

File: src/interfaces/IERC7540.sol [hunk 7]
```diff
@@ -64,12 +111,18 @@ interface IERC7540Redeem {
...
  * @dev    Interface of the ERC7540 "Asynchronous Tokenized Vault Standard", as defined in
  *         https://github.com/ethereum/EIPs/blob/2e63f2096b0c7d8388458bb0a03a7ce0eb3422a4/EIPS/eip-7540.md[ERC-7540].
  */
-interface IERC7540 is IERC7540Deposit, IERC7540Redeem, IERC4626, IERC165 {}
\ No newline at end of file
+interface IERC7540 is IERC7540Deposit, IERC7540Redeem, IERC4626, IERC165 {
+    /**
+     * @dev returns the owner of a given request by requestId
+     * Note: the rid MUST be unique for any given caller but not necessarily any given request. I.e. Rid may be an alias for `owner` OR it may uniquely identify a request. Contracts must be able to handle both cases.
+     *
```
Comment (hieronx) [inline src/interfaces/IERC7540.sol:126]:
```suggestion
     */
```

File: src/interfaces/IERC7540.sol [hunk 8]
```diff
@@ -64,12 +111,18 @@ interface IERC7540Redeem {
...
  *         https://github.com/ethereum/EIPs/blob/2e63f2096b0c7d8388458bb0a03a7ce0eb3422a4/EIPS/eip-7540.md[ERC-7540].
  */
-interface IERC7540 is IERC7540Deposit, IERC7540Redeem, IERC4626, IERC165 {}
\ No newline at end of file
+interface IERC7540 is IERC7540Deposit, IERC7540Redeem, IERC4626, IERC165 {
+    /**
+     * @dev returns the owner of a given request by requestId
+     * Note: the rid MUST be unique for any given caller but not necessarily any given request. I.e. Rid may be an alias for `owner` OR it may uniquely identify a request. Contracts must be able to handle both cases.
+     * 
+    function ownerOf(uint256 rid);
```
Comment (hieronx) [inline src/interfaces/IERC7540.sol:127]:
```suggestion
    function ownerOf(uint256 rid) external view returns (address owner);
```

File: src/interfaces/IERC7540.sol [hunk 9]
```diff
@@ -22,54 +22,144 @@ interface IERC165 {
...
+     * The ERC7540 smart contract calls this function on the recipient
+     * after a `requestRedeem`. This function MAY throw to revert and reject the
+     * request. Return of other than the magic value MUST result in the
+     * transaction being reverted.
+     * Inspired by https://eips.ethereum.org/EIPS/eip-721
+     *
+     * Note: the contract address is always the message sender.
+     *
+     * @param _operator The address which called `requestRedeem` function
+     * @param _sender The address which funded the `shares` of the Request (or message sender)
```
Comment (hieronx) [inline src/interfaces/IERC7540.sol:62]:
```suggestion
     * @param _owner The address which funded the `shares` of the Request (or message sender)
```

File: src/interfaces/IERC7540.sol [hunk 10]
```diff
@@ -22,54 +22,144 @@ interface IERC165 {
...
      * NOTE: most implementations will require pre-approval of the Vault with the Vault's underlying asset token.
+     *
+     * If data is nonzero, attempt to call the receiver onERC7540DepositReceived,
+     * otherwise just send the request to the receiver
      */
-    function requestDeposit(uint256 assets, address operator) external;
-
+    function requestDeposit(uint256 assets, address receiver, address owner, bytes calldata data)
+        external
+        returns (uint256 rid);
```
Comment (hieronx) [inline src/interfaces/IERC7540.sol:96]:
```suggestion
        returns (uint256 requestId);
```

File: src/interfaces/IERC7540.sol [hunk 11]
```diff
@@ -22,54 +22,144 @@ interface IERC165 {
...
+    function pendingDepositRequest(uint256 rid, address owner) external view returns (uint256 pendingAssets);
+
+    /**
+     * @dev Returns the amount of requested assets in Claimable state for the owner to deposit or mint.
+     *
+     * - MUST NOT include any assets in Pending state for deposit or mint.
+     * - MUST NOT show any variations depending on the caller.
+     * - MUST NOT revert unless due to integer overflow caused by an unreasonably large input.
+     */
+    function claimableDepositRequest(uint256 rid, address owner) external view returns (uint256 claimableAssets);
```
Comment (hieronx) [inline src/interfaces/IERC7540.sol:114]:
```suggestion
    function claimableDepositRequest(uint256 requestId, address owner) external view returns (uint256 claimableAssets);
```

File: src/interfaces/IERC7540.sol [hunk 12]
```diff
@@ -22,54 +22,144 @@ interface IERC165 {
...
 }
 
 interface IERC7540Redeem {
-    event RedeemRequest(address indexed sender, address indexed operator, address indexed owner, uint256 shares);
+    event RedeemRequest(address indexed sender, address indexed receiver, address indexed owner, uint256 shares);
```
Comment (hieronx) [inline src/interfaces/IERC7540.sol:118]:
```suggestion
    event RedeemRequest(address indexed receiver, address indexed owner, uint256 indexed requestId, address sender, uint256 assets);
```

File: src/interfaces/IERC7540.sol [hunk 13]
```diff
@@ -22,54 +22,144 @@ interface IERC165 {
...
+     *
+     * NOTE: most implementations will require pre-approval of the Vault with the Vault's share token.
+     *
+     * If data is nonzero, attempt to call the receiver onERC7540RedeemReceived,
+     * otherwise just send the request to the receiver
      */
-    function requestRedeem(uint256 shares, address operator, address owner) external;
+    function requestRedeem(uint256 shares, address receiver, address owner, bytes calldata data)
+        external
+        returns (uint256 rid);
```
Comment (hieronx) [inline src/interfaces/IERC7540.sol:139]:
```suggestion
        returns (uint256 requestId);
```

File: src/interfaces/IERC7540.sol [hunk 14]
```diff
@@ -22,54 +22,144 @@ interface IERC165 {
...
     /**
-     * @dev Returns the amount of requested shares in Pending state for the operator to redeem or withdraw.
+     * @dev Returns the amount of requested shares in Pending state.
      *
      * - MUST NOT include any shares in Claimable state for redeem or withdraw.
      * - MUST NOT show any variations depending on the caller.
      * - MUST NOT revert unless due to integer overflow caused by an unreasonably large input.
      */
-    function pendingRedeemRequest(address operator) external view returns (uint256 shares);
+    function pendingRedeemRequest(uint256 rid, address owner) external view returns (uint256 pendingShares);
```
Comment (hieronx) [inline src/interfaces/IERC7540.sol:148]:
```suggestion
    function pendingRedeemRequest(uint256 requestId, address owner) external view returns (uint256 pendingShares);
```

File: src/interfaces/IERC7540.sol [hunk 15]
```diff
@@ -22,54 +22,144 @@ interface IERC165 {
...
+
+    /**
+     * @dev Returns the amount of requested shares in Claimable state for the owner to redeem or withdraw.
+     *
+     * - MUST NOT include any shares in Pending state for redeem or withdraw.
+     * - MUST NOT show any variations depending on the caller.
+     * - MUST NOT revert unless due to integer overflow caused by an unreasonably large input.
+     */
+    function claimableRedeemRequest(uint256 rid, address owner) external view returns (uint256 claimableShares);
```
Comment (hieronx) [inline src/interfaces/IERC7540.sol:157]:
```suggestion
    function claimableRedeemRequest(uint256 requestId, address owner) external view returns (uint256 claimableShares);
```

**Pull request merged**